### ISSUE TITLE
Context "leaking" between Flask apps

### ISSUE THREAD

**Issue by ibushong** [getsentry/sentry-python#216]
(After a few exchanges with support, I was told to post my issue here and tag @untitaker)

Our application serves multiple flask apps via [subdomain dispatching](http://flask.pocoo.org/docs/1.0/patterns/appdispatch/#dispatch-by-subdomain).

I am noticing that the context/tags set in one app are occasionally "leaking" over into the others, as if the sentry context is not always isolated between apps. Hopefully I am just configuring something wrong, but I don't believe our set-up is especially complicated. 

A brief example/summary: One Flask app runs our API routes/logic and another runs our web admin interface. Both apps have `before_request` hooks that set a Sentry tag called "interface" to either `"api"` or `"admin"`. In general, all issues/errors triggered in the API app show up in Sentry with the correct "interface" tag value (`api`), and same with the admin issues. However, occasionally, there are, for example, issues triggered in the admin app that appear in Sentry tagged with `"api"`. Additionally, the `transaction` value is completely mismatched with the `url` value (for that given event).

Any guidance here? Let me know what other details I can provide to help debug. I can provide links to relevant Sentry issues/events in private communication.

**Comment by untitaker** [comment]
Are you using gevent/eventlet with gunicorn? We don't support that at all.

**Comment by ibushong** [comment]
No, just plain Apache/mod_wsgi (AWS Elastic Beanstalk Python env)

**Comment by ibushong** [comment]
Note that we have set `WSGIApplicationGroup %{GLOBAL}`. Not sure how relevant that might be

**Comment by untitaker** [comment]
Ah sorry, yes I remember this support ticket and the conversation I had with Megan. I have no clue what causes this.

Could you try wrapping the body of your `SubdomainDispatcher.__call__` method with this?

```
from sentry_sdk import Hub

with Hub(Hub.current):
    app = ...
    return app(...)
```

To be clear, this is a shot in the dark.

**Comment by ibushong** [comment]
Yeah I will give that a shot. Also of note, we are applying the [werkzeug.contrib.fixers.ProxyFix](http://flask.pocoo.org/docs/1.0/deploying/wsgi-standalone/#proxy-setups) middleware to each Flask app before passing to the dispatcher. I now realize our dispatching/middleware set-up might be the culprit so here are the guts of that (only included two apps for simplicity):

```python
from werkzeug.contrib.fixers import ProxyFix
from werkzeug.wsgi import pop_path_info, peek_path_info
from sentry_sdk import Hub


class SubdomainDispatcher(object):
    def __init__(self, api_app, admin_app):
        self.api_app = api_app
        self.admin_app = admin_app

    def __call__(self, environ, start_response):
        with Hub(Hub.current):  # Just added this from your suggestion
            host = environ['HTTP_HOST']
            subdomain = host.split(".")[0]

            if subdomain == "api":
                app = self.api_app
            elif subdomain == "admin":
                app = self.admin_app
            else:
                prefix = peek_path_info(environ)
                if prefix == 'api':
                    pop_path_info(environ)  # pop "/api"
                    app = self.api_app
                else:
                    # Default to API
                    app = self.api_app

            return app(environ, start_response)

def create_app():
    from discotech import admin, api

    api_app = api.create_app()
    api_app.wsgi_app = ProxyFix(api_app.wsgi_app)

    admin_app = admin.create_app()
    admin_app.wsgi_app = ProxyFix(admin_app.wsgi_app)

    app = SubdomainDispatcher(api_app, admin_app)

    return app
```

**Comment by untitaker** [comment]
I am pretty sure you need to do `app = ProxyFix(SubDomainDispatcher(app))` instead of applying proxyfix that late. I would try that out first without applying the `with Hub(..)` patch

**Comment by ibushong** [comment]
Yeah that makes more sense. I will give that a try.

**Comment by ibushong** [comment]
Unfortunately, I got the same results after cleaning up the `ProxyFix` implementation (but it's nice to have that cleaned up anyway).

I have implemented `with Hub(Hub.current):` and so far it appears to have fixed the issue! I will continue monitoring and report back tomorrow.

**Comment by untitaker** [comment]
It appears to be sufficient but it's unclear why. Could you run without the fix and enable debug=True on init? 0.6.2 prints out some warnings with that logger that might help

**Comment by ibushong** [comment]
Here is the output from the debug=True logging (init messages, standard "Sending error event" messages, and then there are a few of those internal "unexpected status code: 403" errors)

```
[sentry] DEBUG: Setting up integrations (with default = True)
[sentry] DEBUG: Setting up previously not enabled integration logging
[sentry] DEBUG: Setting up previously not enabled integration flask
[sentry] DEBUG: Setting up previously not enabled integration modules
[sentry] DEBUG: Setting up previously not enabled integration argv
[sentry] DEBUG: Setting up previously not enabled integration celery
[sentry] DEBUG: Setting up previously not enabled integration atexit
[sentry] DEBUG: Setting up previously not enabled integration stdlib
[sentry] DEBUG: Setting up previously not enabled integration excepthook
[sentry] DEBUG: Setting up previously not enabled integration dedupe
[sentry] DEBUG: Enabling integration logging
[sentry] DEBUG: Enabling integration flask
[sentry] DEBUG: Enabling integration modules
[sentry] DEBUG: Enabling integration argv
[sentry] DEBUG: Enabling integration celery
[sentry] DEBUG: Enabling integration atexit
[sentry] DEBUG: Enabling integration stdlib
[sentry] DEBUG: Enabling integration excepthook
[sentry] DEBUG: Enabling integration dedupe
[sentry] DEBUG: Setting up integrations (with default = True)
[sentry] DEBUG: Enabling integration logging
[sentry] DEBUG: Enabling integration flask
[sentry] DEBUG: Enabling integration modules
[sentry] DEBUG: Enabling integration argv
[sentry] DEBUG: Enabling integration celery
[sentry] DEBUG: Enabling integration atexit
[sentry] DEBUG: Enabling integration stdlib
[sentry] DEBUG: Enabling integration excepthook
[sentry] DEBUG: Enabling integration dedupe
[sentry] DEBUG: Killing HTTP transport
[sentry] DEBUG: background worker got kill request
[sentry] DEBUG: Setting up integrations (with default = True)
[sentry] DEBUG: Enabling integration logging
[sentry] DEBUG: Enabling integration flask
[sentry] DEBUG: Enabling integration modules
[sentry] DEBUG: Enabling integration argv
[sentry] DEBUG: Enabling integration celery
[sentry] DEBUG: Enabling integration atexit
[sentry] DEBUG: Enabling integration stdlib
[sentry] DEBUG: Enabling integration excepthook
[sentry] DEBUG: Enabling integration dedupe
[sentry] DEBUG: Killing HTTP transport
[sentry] DEBUG: background worker got kill request
[sentry] DEBUG: Setting up integrations (with default = True)
[sentry] DEBUG: Enabling integration logging
[sentry] DEBUG: Enabling integration flask
[sentry] DEBUG: Enabling integration modules
[sentry] DEBUG: Enabling integration argv
[sentry] DEBUG: Enabling integration celery
[sentry] DEBUG: Enabling integration atexit
[sentry] DEBUG: Enabling integration stdlib
[sentry] DEBUG: Enabling integration excepthook
[sentry] DEBUG: Enabling integration dedupe
[sentry] DEBUG: Killing HTTP transport
[sentry] DEBUG: background worker got kill request
[sentry] DEBUG: Setting up integrations (with default = True)
[sentry] DEBUG: Enabling integration logging
[sentry] DEBUG: Enabling integration flask
[sentry] DEBUG: Enabling integration modules
[sentry] DEBUG: Enabling integration argv
[sentry] DEBUG: Enabling integration celery
[sentry] DEBUG: Enabling integration atexit
[sentry] DEBUG: Enabling integration stdlib
[sentry] DEBUG: Enabling integration excepthook
[sentry] DEBUG: Enabling integration dedupe
[sentry] DEBUG: Killing HTTP transport
[sentry] DEBUG: background worker got kill request

----SNIP----

[sentry] DEBUG: Sending error event [a1c90484ff914ae89e7646c0b81a29e4] to app.getsentry.com project:82926

----SNIP----

 [sentry] DEBUG: Sending error event [75c1516769a14418a5937de9cf84456e] to app.getsentry.com project:82926
 [sentry] ERROR: Internal error in sentry_sdk
Traceback (most recent call last):
  File "/opt/python/run/venv/local/lib/python2.7/site-packages/sentry_sdk/utils.py", line 43, in capture_internal_exceptions
    yield
  File "/opt/python/run/venv/local/lib/python2.7/site-packages/sentry_sdk/transport.py", line 157, in send_event_wrapper
    self._send_event(event)
  File "/opt/python/run/venv/local/lib/python2.7/site-packages/sentry_sdk/transport.py", line 145, in _send_event
    raise ValueError("Unexpected status code: %s" % response.status)
ValueError: Unexpected status code: 403
```

**Comment by untitaker** [comment]
So what I am mostly interested in are log lines with the log level WARNING. I assume you didn't see any of that?

**Comment by untitaker** [comment]
BTW https://github.com/getsentry/sentry-python/pull/218 is a PR that I want to merge but I am not sure yet if it's a good idea. If you have time to try it out that'd be great though

**Comment by ibushong** [comment]
Correct, there are no `WARNING` messages.

Ah ok. I need to spend some time on other things now that I have this in a decent state. I will try to get to that later this week though. Thanks so much for the help.

**Comment by untitaker** [comment]
release 0.6.3 should fix this, please try

**Comment by ibushong** [comment]
Excellent! I will try it out

**Comment by ibushong** [comment]
Ok yeah so far everything looks good using v0.6.3 (and with the previous `Hub` fix removed). Thanks for the super quick turnaround!

**Comment by mitsuhiko** [comment]
Closing this as resolved.

**Comment by michaelbukachi** [comment]
@ibushong I'm curious to know how you implemented sentry for each app. Currently doing a similar setup and but we are kinda stuck. We keep getting a 
```
Warning: Integration <class 'sentry_sdk.integrations.stdlib.StdlibIntegration'> attempted to run but it was only enabled on init() but not the client that was bound to the current flow.  Earlier versions of the SDK would consider these integrations enabled but this is no longer the case.
```

**Comment by untitaker** [comment]
@michaelbukachi this was a bug in the SDK and has been resolved in 0.6.3. I am quite certain your issue is unrelated. Please file a new issue or ask on StackOverflow.
